Proposal #11 - Use FHIR as the internal format (#1921)
CDCgov/prime-reportstream: prime-router/docs/proposals/0011-internal-formats/0011-internal-formats.md

# Considering New Internal Formats 

This proposal is part of a series of recommendations that try to solve multiple problems together in one suggestion. These proposals consider the learnings and the experience from operating ReportStream for the past six months and the current production directions for the service for the next six months. 

## Going beyond COVID-19
USDS created ReportStream in response to the COVID-19 public health emergency, but with the thought that it would last beyond this emergency. Although the public health emergency is not over, several product requests indicate that it is time to consider how to take ReportStream beyond the public health emergency. Among these are: 

> 1. Full ELR support for HCA: The HCA team has requested full ELR support from ReportStream first and VXU and ADT support in the future.  

Reviewer: ```suggestion
1. Full ELR support, followed by VXU and ADT support in the future.  
```
```suggestion
1. Full ELR support for HCA: The HCA team has requested full ELR support from ReportStream first and VXU and ADT support in the future.  
```

### Summary
The table below grades each option on a 1-5 scale, where five is best. 

|            | New Use-Cases | New People | Guardrails  | Initial Effort  |
|:----       |:----------:   |:----------:|:-----------:|:---------------:|
| HL7        | 3             | 3          | 2           | 4               |
| FHIR       | 5             | 4          | 5           | 2               |
| Custom     | 4             | 2          | 1           | 5               |

## Recommendation
Although an HL7 internal format would be the quickest path to support full ELR, since ELR is an HL7-based standard, this path would not support the future FHIR scenarios. HL7 also has problems representing new data elements like those that SimpleReport uses to capture new testing sites. Although it may take more initial work, ReportStream should use FHIR as its internal structure because it supports more future scenarios and benefits engineers.  

## Details

### Topics

> In this proposal, the concept of a topic becomes something like parts of a FHIR implementation guide which defines what FHIR resources are in a particular bundle. Further investigation should be done to see if they are the same concept. 

Reviewer: We definitely need _something_ to provide a clear description of what we expect from a sender - and right now that's our .schema files.  So, yes, let's examine whether FHIR profiles could replace our current schemas is well worth considering.   This can be treated as a separate follow-on project to the FHIR-in-the-hole project.

Despite being designed for FHIR, the FHIR profiles would have to "work" for CSVs too - CSVs are not going away.

### Data elements

Similar to the XPath concept for XML, FHIR defines a path syntax for its data elements. The FHIR path replaces the data element name in most cases in the current engine. The following table shows how some of the current data elements will translate into FHIR paths. 

| Current Name          | FHIR Path   | Comments | 
|:---------|:----------|:-----|
| patient_last_name     | Patient.name.family   | FHIR follows USCDI naming conventions  | 
| patient_first_name    | Patient.name.given[0] | FHIR supports multiple given names |
| specimen_source_site  | Specimen.collection.bodySite | SNOMED-CT value |
| ordering_provider_npi | ServiceRequest.requester.identifier | NPI value with CMS as the assigner |
| site_of_care | ServiceRequest.location.extension.siteOfCare | Site of care is an extension on the location of the request |

> 

Reviewer: Definitely both a syntactic and semantic improvement.

### Proposal 10 - Translations and Schemas
This proposal can be thought of as an extension to proposal 10, which reduces the role of a schema. This proposal says the internal format is the FHIR standard implementation guide for a particular topic, thus eliminating the schema altogether. 

### Mappers and transforms

> In combination with proposal #10, the Mappers types can be replaced by whole message transforms. 

Reviewer: I think @whytheplatypus is working on a proposal that elucidates a path forwards on this, close to what you've outlined in this and the other Proposal.
